cellmlmanip/units.py

```diff
@@ -162,146 +143,145 @@ def __init__(self, expression):
+        """
+        assert name not in CELLML_UNITS, 'Cannot redefine CellML unit <%s>' % name
+        assert not self.is_unit_defined(name), 'Unit <%s> already exists' % name
 
-        if unit_definition is not None:
-            assert not self._is_unit_defined(units_name), 'Unit <%s> already exists' % units_name
+        # TODO ADD NAME PREFIX
 
-            # check whether the definition is dimensionless (e.g. a dimensionless scaling factor)
-            definition = self.ureg.parse_expression(unit_definition.split('=')[1])
-            # if the unit is dimensionless
-            if definition.dimensionless:
-                # get the scale factor by converting the quantity to dimensionless
-                scale_factor = definition.to(self.ureg.dimensionless)
-                # dimensionless units can't be created using definition strings
-                unit_definition = UnitDefinition(units_name, '', (), ScaleConverter(scale_factor))
+        # Dimensionless units can't be created using a string expression.
+        # To test if this is a dimensionless unit, parse the string as a Quantity and check if it's dimensionless
+        quantity = self._registry.parse_expression(expression)
+        if quantity.units == self._registry.dimensionless:
+            definition = UnitDefinition(name, '', (), ScaleConverter(quantity.to(self._registry.dimensionless)))
+        else:
+            definition = expression
 
-            self._define_pint_unit(units_name, unit_definition)
+        # Add to registry and list of custom units
+        self._registry.define(name + '=' + definition)
+        self._custom_units.add(name)
 
-    def add_base_unit(self, units_name):
+    def add_base_unit(self, name):
         """Define a new base unit in the Pint registry.
 
         :param units_name: string name of unit to add
         """
-        assert units_name not in CELLML_UNITS, 'Cannot redefine CellML unit <%s>' % units_name
-        assert not self._is_unit_defined(units_name), 'Unit <%s> already exists' % units_name
-        self._define_pint_unit(units_name, '{name}=[{name}]'.format_map({'name': units_name}))
+        assert name not in CELLML_UNITS, 'Cannot redefine CellML unit <%s>' % name
+        assert not self.is_unit_defined(name), 'Unit <%s> already exists' % name
+
+        # TODO ADD NAME PREFIX
+
+        # Add to registry and list of custom units
+        self._registry.define(name + '=[' + name + ']')
+        self._custom_units.add(name)
 
-    def _is_unit_defined(self, name):
-        """ Check whether the unit already exists in Pint registry.
+    def is_unit_defined(self, name):
+        """Check if a unit with the given ``name`` exists.
 
         :param name: string name of the unit
         :returns: True if exists, else False
         """
+        # TODO ADD PREFIX
+
         try:
-            getattr(self.ureg, name)
+            getattr(self._registry, name)
```

Comment:
I was thinking we could use (a variant of) `self._custom_units` for this, if it also contained the built-in CellML names?